### DISCUSSION TITLE
Configurable JSON backend

### DISCUSSION THREAD

**Discussion by kirillkaiumov** [lostisland/faraday#1454] [Help]
Currently `faraday` provides a [middleware](https://github.com/lostisland/faraday/blob/36916f07596f6e1ab71b688be56e5e087a1372df/lib/faraday/response/json.rb) to encode/decode JSON payloads. Internally it uses `json` that from the Ruby's standard library. If I want to use another JSON backend like [oj](https://github.com/ohler55/oj) then I need to implement my own middleware. So, I have a few questions:
1. Do you think it's a good idea to allow configuring the built-in middleware to work with different JSON backends?
2. If I want to make my own middleware to use `oj`, can I make it inherit from `Faraday::Response::Json` class? Or it's considered "private" and better to inherit from `Faraday::Middleware`?

Thanks!

**Comment by olleolleolle** [comment]
👋 I turned this from an Issue into a Discussion, as this seemed more appropriate. Hope this helps!

**Comment by yykamei** [comment]
I think it's OK to inherit another middleware as Faraday Multipart does.

https://github.com/lostisland/faraday-multipart/blob/a3609a6921cd032252a9cee5d1388982e54bc045/lib/faraday/multipart/middleware.rb#L8

Also, I think it makes sense with the first solution. For example, Faraday's built-in Logger middleware has an option to receive a user-specified logger here.

https://github.com/lostisland/faraday/blob/36916f07596f6e1ab71b688be56e5e087a1372df/lib/faraday/response/logger.rb#L15

I prefer the first one, but it seems to require a lot of work for it.

**Reply by iMacTia** [reply]
Thank you @yykamei, this pretty much resonates with my view as well 🙌!

**Comment by iMacTia** [comment]
I just want to add to @yykamei [answer](https://github.com/lostisland/faraday/discussions/1454#discussioncomment-3869491) a few thoughts at the maintainer:
* `json` is a standard gem, and thus we know it's always available. Using `oj` would require adding it as a dependency, or have some weird runtime check like `if defined?(Oj) ... else ...`. Faraday 2.0 was all about moving away from external dependencies, so I'd much rather prefer a specialised middleware instead.
* It's perfectly fine to inherit from `Faraday::Response::Json` and override the `parse` method. If the fact that the method is private is causing issues, we can make it public/protected instead

**Reply by iMacTia** [reply]
FYI -- a [gem already exists](https://github.com/7even/faraday_middleware-parse_oj/blob/master/lib/faraday_middleware/parse_oj.rb) to do this, but it's based off the old JSON middleware included with `faraday_middleware`, so it will probably need an update, or you can use that as inspiration and build a new one 😄!

If you do, feel free to add it to [awesome-faraday](https://github.com/lostisland/awesome-faraday) so people know it exists 🙌

**Comment by iMacTia** [comment]
@kirillkaiumov I'm happy to say a contribution was just merged that makes JSON encoders/decoders configurable, which I believe is what you were after 😄 
https://github.com/lostisland/faraday/pull/1539

I'll release it soon in a new minor version 👍
